File: android/src/main/kotlin/africa/ejara/trustdart/coins/DOGE.kt
import africa.ejara.trustdart.Coin
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toHexByteArray
import africa.ejara.trustdart.utils.toHexBytesInByteString
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import wallet.core.jni.BitcoinAddress
import wallet.core.jni.BitcoinScript
import wallet.core.jni.BitcoinSigHashType
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Bitcoin

class DOGE : Coin("DOGE", CoinType.DOGECOIN ) {

     override fun signTransaction(
         path: String,
         txData: Map<String, Any>,
         mnemonic: String,
         passphrase: String
     ): String? {
         val wallet = HDWallet(mnemonic, passphrase)
         val privateKey = wallet.getKey(coinType, path)
         val publicKey = privateKey.getPublicKeySecp256k1(true)
         val address = BitcoinAddress(publicKey, coinType!!.p2pkhPrefix())
         val utxos: List<Map<String, Any>> = txData["utxos"] as List<Map<String, Any>>

         val script = BitcoinScript.lockScriptForAddress(address.description(), coinType)

         val input = Bitcoin.SigningInput.newBuilder()
         .setAmount((txData["amount"] as Int).toLong())
         .setHashType(BitcoinSigHashType.ALL.value())
         .setToAddress(txData["toAddress"] as String)
         .setChangeAddress(txData["changeAddress"] as String)
         .setByteFee((txData["fees"] as Int).toLong())
         .setCoinType(CoinType.DOGECOIN.value())
         .addPrivateKey(ByteString.copyFrom(privateKey.data()))

    
        for (utx in utxos) {
            val txId = Numeric.hexStringToByteArray(utx["txid"] as String)
            val outPoint = Bitcoin.OutPoint.newBuilder()
                .setHash(ByteString.copyFrom(txId.reversedArray()))
                .setIndex(utx["vout"] as Int)
                .build()

            val utxo = Bitcoin.UnspentTransaction.newBuilder()
                .setAmount(utx["value"] as Long)
                .setOutPoint(outPoint)
                .setScript(ByteString.copyFrom(script.data()))
                .build()

            input.addUtxo(utxo)
        }

        val output = AnySigner.sign(input.build(), coinType, Bitcoin.SigningOutput.parser())

         return Numeric.toHexString(output.encoded.toByteArray())
     }

}

File: ios/Classes/coins/DOGE.swift
/*
 DOGE
 */

import WalletCore

class DOGE: Coin  {
    init() {
        super.init(name: "DOGE", coinType: .dogecoin)
    }

    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: coinType, derivationPath: path)
        let publicKey = privateKey!.getPublicKeySecp256k1(compressed: true)
        let address = BitcoinAddress(publicKey: publicKey, prefix: coinType.p2pkhPrefix)
        let script = BitcoinScript.lockScriptForAddress(address: address!.description, coin: coinType)
        let utxos: [[String: Any]] = txData["utxos"] as! [[String: Any]]
        var unspent: [BitcoinUnspentTransaction] = []
        
        if privateKey != nil {
            for utx in utxos {
                unspent.append(BitcoinUnspentTransaction.with {
                    $0.outPoint.hash = Data.reverse(hexString: utx["txid"] as! String)
                    $0.outPoint.index = UInt32(utx["vout"] as! Int)
                    $0.outPoint.sequence = UINT32_MAX
                    $0.amount = utx["value"] as! Int64
                    $0.script = script.data
                })
            }
            
            
            let input: BitcoinSigningInput = BitcoinSigningInput.with {
                $0.hashType = BitcoinScript.hashTypeForCoin(coinType: coinType)
                $0.toAddress = txData["toAddress"] as! String
                $0.changeAddress = txData["changeAddress"] as! String
                $0.privateKey = [privateKey!.data]
                $0.amount = txData["amount"] as! Int64
                $0.coinType = coinType.rawValue
                $0.byteFee = txData["fees"] as! Int64
                $0.utxo = unspent
            }
            
            let output: BitcoinSigningOutput = AnySigner.sign(input: input, coin: coinType)
            return output.encoded.hexString
        }else {
            return nil
        }
        
    }
}
